core/desugarSwitchFuns.ml

```diff
@@ -0,0 +1,115 @@
+open Sugartypes
+open Utility
+open SourceCode
+
+(* This module desugars pattern-matching functions
+
+  This transformation convert `switch` functions of the form:
+
+  fun foo(a1, ..., an) switch {
+    case (p1_1, ..., p1_n) -> b_1
+    ...
+    case (pm_1, pm_n) -> b_m
+  }
+
+  to standard functions of the form:
+
+  fun foo(a1 as x1, ..., an as xn) {
+    switch ((x1, ..., xn)) {
+      case (p1_1, ..., p1_n) -> b_1
+      ...
+      case (pm_1, ..., pm_n) -> b_m
+      case (_, ..., _) -> error("non-exhaustive")
+  }
+
+  The last non-exhaustive case with wild card pattern is always attached to the end of switch body.
+
+*)
+
+let with_pos = SourceCode.WithPos.make
+
+let pattern_matching_sugar =
+  Settings.(
+    flag "pattern_matching_sugar"
+    |> synopsis
+         "Toggles whether to enable the switch pattern matching syntax sugar"
+    |> convert parse_bool
+    |> sync)
+
+let pattern_matching_sugar_guard pos =
+  let pattern_matching_sugar_disabled pos =
+    Errors.disabled_extension ~pos ~setting:("pattern_matching_sugar", true) "Pattern Matching Sugar"
+  in
+  if not (Settings.get pattern_matching_sugar)
+  then raise (pattern_matching_sugar_disabled pos)
+
+let nullary_guard pss pos =
+  let nullary_error pos =
+    Errors.desugaring_error ~pos:pos ~stage:Errors.DesugarSwitchFuns ~message:"Can't match over nullary function"
+  in
+  match pss with
+    | [] -> raise (nullary_error pos)
+    | _ -> ()
+
+let switch_fun_currying_guard pos args =
+  match args with
+  | [arg] -> arg
+  | _ -> raise (Errors.Type_error (pos, "Curried switch functions are not yet supported."))
+
+let construct_normal_funlit pos patterns cases =
+  pattern_matching_sugar_guard pos;
+  let patterns = switch_fun_currying_guard pos patterns in
+  nullary_guard patterns pos;
+  (* bind the arguments with unique var name *)
+  let name_list = List.map (fun pat -> (pat, Utility.gensym())) patterns in
+  let switch_tuple = List.map (fun (_, name) -> with_pos (Var name)) name_list in
+  (* assemble exhaustive handler *)
+  let exhaustive_patterns = with_pos (Pattern.Any) in
+  let exhaustive_position = Format.sprintf "non-exhaustive pattern matching at %s" (SourceCode.Position.show pos) in
+  let exhaustive_case = FnAppl (with_pos (Var "error"), [with_pos (Constant (CommonTypes.Constant.String exhaustive_position))]) in
+  let normal_args =
+    List.map
+      (fun (pat, name) -> with_pos (Pattern.As (with_pos (Binder.make ~name ()), pat)))
+      name_list
+  in
+  let cases = cases@[(exhaustive_patterns, with_pos exhaustive_case)] in
+  let switch_body = Switch (with_pos (TupleLit switch_tuple), cases, None) in
+  let normal_fnlit = NormalFunlit ([normal_args], with_pos switch_body) in
```

Comment:
I think you should reuse the source position of `fun f(...) switch` for `switch_body` to improve error messages. Right now you get error messages like
```
:0: Type error: The type of an input to a switch should match the type of its patterns, but the expression
    `<dummy>'
```
where `<dummy>`, in this instance, is the `switch_body`.

Reply:
Sorry, forget to check the error message.

Reply:
No worries. They have sort of secondary status anyways -- the average quality of the error messages in Links is pretty poor.

Generally, it is a good idea to map source position from sugared nodes to desugared nodes whenever possible. It may mean that several desugared nodes may share the same position, because they were derived from the same sugared node -- which is fine. At least this provides the user with a minimal degree of provenance.

As an aside: One thing we lack internally is the ability to tell whether a given node is synthetic, i.e. desugared.

Reply:
Fixed. Would you like to give me an example for testing it?

Reply:
Try some that are obvious type errors:
```
fun f(xs) switch { case [] -> xs case x :: _ -> x } # body types do not match
fun g(Foo) switch { case 1 -> 0 } # incompatible scrutinee and pattern types.
```